### PR TITLE
[AMD] Add MoE weights and scales padding

### PR DESCRIPTION
**Pull request by mqhc2020**
## Motivation

Right now, Aiter MoE requires weights and scales to align with a fixed number. Since some models have intermediate sizes that don't fit this rule, we need to add extra padding to the weights so they can be processed by the Fused MoE.

## Modifications

Add padding for the weights. Below listed are the models and configurations that has been verified with:

1. Qwen/Qwen3-235B-A22B-Instruct-2507-FP8 : TP=8
2. amd/Qwen3-235B-A22B-Instruct-2507-MXFP4: TP=8
3. amd/Qwen3.5-397B-A17B-MXFP4: TP=8
4. zai-org/GLM-4.7: TP=8

## Accuracy Tests

We run GSM8K accuracy test for models below:

**Qwen/Qwen3-235B-A22B-Instruct-2507-FP8 (TP8):**

Serve command:
```
SGLANG_USE_AITER=1 python3 -m sglang.launch_server \
	--model-path ${MODEL_PATH} \
	--tp-size 8 \
	--ep-size 1 \
	--trust-remote-code \
	--chunked-prefill-size 131072 \
	--mem-fraction-static 0.9 \
	--attention-backend aiter
```

Accuracy: 0.980
Invalid: 0.000
Latency: 10.525 s
Output throughput: 2663.037 token/s

**amd/Qwen3-235B-A22B-Instruct-2507-MXFP4 (TP8):**

Serve command:
```
SGLANG_USE_AITER=1 python3 -m sglang.launch_server \
	--model-path ${MODEL_PATH} \
	--tp-size 8 \
	--ep-size 1 \
	--trust-remote-code \
	--chunked-prefill-size 131072 \
	--mem-fraction-static 0.9 \
	--attention-backend aiter
```

Accuracy: 0.940
Invalid: 0.000
Latency: 23.853 s
Output throughput: 7488.441 token/s

**zai-org/GLM-4.7 (TP8):**

Serve command:
```
SGLANG_USE_AITER=1 python3 -m sglang.launch_server \
	--model $MODEL_PATH \
	--tp 8 \
	--tool-call-parser glm47 \
	--reasoning-parser glm45
```

(better than VLLM TP4 on model card)

Accuracy: 0.975
Invalid: 0.000
Latency: 7.349 s
Output throughput: 2683.189 token/s

Serving command:
`python3 ./benchmark/gsm8k/bench_sglang.py --num-questions 1319 --parallel 1319`

## Benchmarking and Profiling

Baseline benchmark for amd/Qwen3-235B-A22B-Instruct-2507-mxfp4 (TP4) and Qwen/Qwen3-235B-A22B-Instruct-2507-FP8 (TP8):
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
5. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
6. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
7. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py [hunk 1]
```diff
@@ -112,17 +123,24 @@ def create_weights(
...
                 dtype=params_dtype,
             ),
             requires_grad=False,
         )
+
+        W2_SCALE_DIVIDEND = w2_down_dim * 2
+        W2_SCALE_DIVISOR = intermediate_size_per_partition
+        scaling_up = lambda dividend, divisor: (dividend * W2_SCALE_DIVIDEND) // (
+            divisor * W2_SCALE_DIVISOR
+        )
```
[the remark this answers is not in the document]
Comment (mqhc2020) [reply python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py]: Simplify the logic and add comments, so it's now easier to understand.

File: python/sglang/srt/layers/moe/utils.py [hunk 2]
```diff
@@ -341,3 +344,30 @@ class RoutingMethodType(IntEnum):
...
+    else:
+        return 128 if bool(int(os.getenv("SGLANG_MOE_PADDING", "0"))) else 0
+
+
+def get_moe_weight_sizes(inter_dim, is_concat, is_packed, is_aiter_moe):
+    w13_up_dim = 2 * inter_dim if is_concat else inter_dim
+    w2_down_dim = inter_dim // 2 if is_packed else inter_dim
+
+    if is_aiter_moe:
+        padding_size = get_moe_padding_size(True)
```
[the remark this answers is not in the document]
Comment (mqhc2020) [reply python/sglang/srt/layers/moe/utils.py:390]: Done.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 3]
```diff
@@ -514,8 +520,14 @@ def _load_w2(
...
         # 2. GPU with flashinfer_trtllm padding (when intermediate_size is padded to 128)
+        # 3. GPU with Aiter padding
         # This handles the case where the loaded weights are smaller than the padded expert_data
-        use_padded_loading = _is_cpu or self.use_flashinfer_trtllm_moe
+        aiter_padded = (
+            _use_aiter
+            and hasattr(self, "w2_weight")
+            and getattr(self.w2_weight, "weight_padded", False)
+        )
+        use_padded_loading = _is_cpu or self.use_flashinfer_trtllm_moe or aiter_padded
```
[the remark this answers is not in the document]
Comment (mqhc2020) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: It cannot be a property because the attributes are decided much later, but a helper method is add.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 4]
```diff
@@ -316,6 +316,20 @@ def __init__(
         if self.quant_method is not None and hasattr(self.quant_method, "runner"):
             self.runner = self.quant_method.runner
 
+    def use_padded_loading(self) -> bool:
```
Comment (hubertlu-tw) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py:330]: Maybe change it to avoid repeated attribute lookups?

@functools.cached_property
def _use_padded_loading(self) -> bool:

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 5]
```diff
@@ -427,12 +441,7 @@ def _load_w13(
         else:
             start = 0
 
-        # Use narrow_padded_param_and_loaded_weight for:
-        # 1. CPU (always)
-        # 2. GPU with flashinfer_trtllm padding (when intermediate_size is padded to 128)
-        # This handles the case where the loaded weights are smaller than the padded expert_data
-        use_padded_loading = _is_cpu or self.use_flashinfer_trtllm_moe
-        if use_padded_loading:
+        if self.use_padded_loading():
```
Comment (hubertlu-tw) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Maybe change to `if self._use_padded_loading:` accordingly?

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 6]
```diff
@@ -501,12 +510,7 @@ def _load_w2(
             # for w2 in TP, it shards the input_features, i.e., shard_dim=2
             shard_size = expert_data.shape[shard_dim]
 
-        # Use narrow_padded_param_and_loaded_weight for:
-        # 1. CPU (always)
-        # 2. GPU with flashinfer_trtllm padding (when intermediate_size is padded to 128)
-        # This handles the case where the loaded weights are smaller than the padded expert_data
-        use_padded_loading = _is_cpu or self.use_flashinfer_trtllm_moe
-        if use_padded_loading:
+        if self.use_padded_loading():
```
Comment (hubertlu-tw) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Maybe change to `if self._use_padded_loading:` accordingly?

**Review by hubertlu-tw — CHANGES_REQUESTED** [review]
Other than a few comments I left, LGTM. Notice that this PR will also resolve an issue reported here:
https://github.com/sgl-project/sglang/issues/21918#issuecomment-4180191842

CC: @HaiShaw

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 4 above)
Comment (mqhc2020) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py:330]: I tried to add it as a property in the first place but failed due to the use of globals (instead of class attributes). Maybe cached_property is doable.

File: python/sglang/srt/layers/moe/utils.py [hunk 7]
```diff
@@ -341,3 +344,38 @@ class RoutingMethodType(IntEnum):
...
+        return AITER_PADDING_SIZE
+    else:
+        return (
+            TRITON_PADDING_SIZE
+            if bool(int(os.getenv("SGLANG_MOE_PADDING", "0")))
+            else 0
+        )
+
+
+def get_moe_weight_sizes(inter_dim, is_concat, is_packed, is_aiter_moe):
```
Comment (kkHuang-amd) [inline python/sglang/srt/layers/moe/utils.py:375]: The meaning of is_packed is not clear.

From your logic, it means the quantization format is 4bits quantization.

You can have more clear name to identify this variable or add some comment to explain it.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 4 above)
Comment (mqhc2020) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py:330]: Modified accordingly in [347a42f](https://github.com/sgl-project/sglang/pull/21097/commits/347a42f4e9bcb88469f7c1d77ddbd015c24472f2). That's very helpful, thanks a lot.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 5 above)
Comment (mqhc2020) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Changed accordingly.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 6 above)
Comment (mqhc2020) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Changed accordingly.

**Review by HaiShaw — COMMENTED** [review]
@mqhc2020 please run a verification on grok2 and Grok INT4-FP8 MoE model.

**Comment by mqhc2020** [comment]
> @mqhc2020 could you please also add the commands you ran on the PR description? It may help the review process. Thanks!

Here are the accuracy results:
 
**Grok1 INT4-FP8**
 
Accuracy: 0.825
Invalid: 0.000
Latency: 141.465 s
Output throughput: 1338.350 token/s
 
**Grok2**
 
Accuracy: 0.929
Invalid: 0.000
Latency: 177.871 s
Output throughput: 840.041 token/s
 
Note that:

1. Also checked that the MoE path for grok-1 and grok-2 is not affected (not block quant)
2. There was a problem related to pre-shared models and that was considered before.

**Review by hubertlu-tw — APPROVED** [review]
LGTM

**Comment by hubertlu-tw** [comment]
@HaiShaw could you please review this PR again? This will resolve an issue reported by https://github.com/sgl-project/sglang/issues/21918

File: python/sglang/srt/layers/moe/utils.py [hunk 8]
```diff
@@ -351,3 +354,47 @@ class RoutingMethodType(IntEnum):
...
+        align_aiter = lambda n: ((n + padding_size - 1) // padding_size) * padding_size
+        is_padded = (w2_down_dim % padding_size) > 0
+        if is_padded:
+            w2_down_dim = align_aiter(w2_down_dim)
+        # up proj + gate fusion : 2x
+        if is_concat:
+            w13_up_dim = w2_down_dim * 2
+        # packed
+        if hasattr(torch, "float4_e2m1fn_x2") and is_packed:
+            w13_up_dim *= 2
```
Comment (HaiShaw) [inline python/sglang/srt/layers/moe/utils.py:402]: why is this `*2` needed?

**Review by HaiShaw — APPROVED** [review]
